esquery.js

```diff
@@ -337,30 +407,27 @@ function adjacent(node, selector, ancestry, side, options) {
     return false;
 }
 
-/**
-* @callback IndexFunction
-* @param {Integer} len Containing list's length
-* @returns {Integer}
-*/
-
 /**
  * Determines if the given node is the nth child, determined by
  * `idxFn`, which is given the containing list's length.
  * @param {external:AST} node
  * @param {external:AST[]} ancestry
- * @param {IndexFunction} idxFn
+ * @param {Integer} nth
+ * @param {boolean} fromEnd
  * @param {ESQueryOptions|undefined} options
  * @returns {boolean}
  */
-function nthChild(node, ancestry, idxFn, options) {
+function nthChild(node, ancestry, nth, fromEnd, options) {
+    if (nth <= 0) { return false; }
     const [parent] = ancestry;
     if (!parent) { return false; }
     const keys = getVisitorKeys(parent, options);
-    for (const key of keys) {
-        const listProp = parent[key];
-        if (Array.isArray(listProp)) {
-            const idx = listProp.indexOf(node);
-            if (idx >= 0 && idx === idxFn(listProp.length)) { return true; }
+    for (let i = 0; i < keys.length; i++) {
+        const listProp = parent[keys[i]];
+        if (Array.isArray(listProp)){
+            if (nth <= listProp.length && listProp[fromEnd ? listProp.length-nth : nth-1] === node) {
```

Comment:
Style nit: spaces around infix operators.

```suggestion
            if (nth <= listProp.length && listProp[fromEnd ? listProp.length - nth : nth - 1] === node) {
```